src/pool/pool_tbb.c

```diff
@@ -0,0 +1,200 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <dlfcn.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <pthread.h>
+
+#include <umf/memory_pool.h>
+#include <umf/memory_pool_ops.h>
+#include <umf/memory_provider.h>
+
+#include "umf/pools/pool_tbb.h"
+
+typedef void *(*raw_alloc_tbb_type)(intptr_t, size_t *);
+typedef void (*raw_free_tbb_type)(intptr_t, void *, size_t);
+
+struct mem_pool_policy_s {
+    raw_alloc_tbb_type pAlloc;
+    raw_free_tbb_type pFree;
+    size_t granularity;
+    int version;
+    unsigned fixed_pool : 1, keep_all_memory : 1, reserved : 30;
+};
+
+struct tbb_callbacks {
+    void *(*pool_malloc)(void *, size_t);
+    void *(*pool_realloc)(void *, void *, size_t);
+    void *(*pool_aligned_malloc)(void *, size_t, size_t);
+    bool (*pool_free)(void *, void *);
+    int (*pool_create_v1)(intptr_t, const struct mem_pool_policy_s *, void **);
+    bool (*pool_destroy)(void *);
+    void *(*pool_identify)(void *object);
+    size_t (*pool_msize)(void *, void *);
+};
+
+struct tbb_memory_pool {
+    umf_memory_provider_handle_t mem_provider;
+    void *tbb_pool;
+};
+
+static struct tbb_callbacks g_tbb_ops;
+static pthread_once_t tbb_is_initialized = PTHREAD_ONCE_INIT;
+
+void load_tbb_symbols(void) {
+    fprintf(stdout, "Initializing TBB symbols\n");
+    const char so_name[] = "libtbbmalloc.so.2";
+    void *tbb_handle = dlopen(so_name, RTLD_LAZY);
+    if (!tbb_handle) {
+        fprintf(stderr, "%s not found.\n", so_name);
+        abort();
+    }
+
+    struct tbb_callbacks tbb_ops;
+
+    *(void **)&tbb_ops.pool_malloc =
+        dlsym(tbb_handle, "_ZN3rml11pool_mallocEPNS_10MemoryPoolEm");
+    *(void **)&tbb_ops.pool_realloc =
+        dlsym(tbb_handle, "_ZN3rml12pool_reallocEPNS_10MemoryPoolEPvm");
+    *(void **)&tbb_ops.pool_aligned_malloc =
+        dlsym(tbb_handle, "_ZN3rml19pool_aligned_mallocEPNS_10MemoryPoolEmm");
+    *(void **)&tbb_ops.pool_free =
+        dlsym(tbb_handle, "_ZN3rml9pool_freeEPNS_10MemoryPoolEPv");
+    *(void **)&tbb_ops.pool_create_v1 = dlsym(
+        tbb_handle,
+        "_ZN3rml14pool_create_v1ElPKNS_13MemPoolPolicyEPPNS_10MemoryPoolE");
+    *(void **)&tbb_ops.pool_destroy =
+        dlsym(tbb_handle, "_ZN3rml12pool_destroyEPNS_10MemoryPoolE");
+    *(void **)&tbb_ops.pool_identify =
+        dlsym(tbb_handle, "_ZN3rml13pool_identifyEPv");
+    *(void **)&tbb_ops.pool_msize =
+        dlsym(tbb_handle, "_ZN3rml10pool_msizeEPNS_10MemoryPoolEPv");
+
+    if (!tbb_ops.pool_malloc || !tbb_ops.pool_realloc ||
+        !tbb_ops.pool_aligned_malloc || !tbb_ops.pool_free ||
+        !tbb_ops.pool_create_v1 || !tbb_ops.pool_destroy ||
+        !tbb_ops.pool_identify) {
+        fprintf(stderr, "Could not find symbols in %s.\n", so_name);
+        dlclose(tbb_handle);
+        abort();
```

Comment:
Return an error.

Reply:
Cannot - this function has to return `void`, because it is an argument in `pthread_once()`

Reply:
Same as above.

Reply:
Right. Done.